extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts

```diff
@@ -190,92 +167,105 @@ export class MssqlObjectExplorerNodeProvider extends ProviderBase implements sql
-				return session;
-			}
-		}
-		return undefined;
+	public findSqlClusterSessionBySqlConnProfile(sqlConnProfile: sqlops.IConnectionProfile): SqlClusterSession {
+		return Array.from(this.sessionMap.values())
+			.find(s => s.isMatchedSqlConnProfile(sqlConnProfile));
 	}
 }
 
-export class Session {
-	private _root: RootNode;
-	constructor(private _connection: Connection, private sessionId?: string) {
-	}
+export class SqlClusterSession {
+	private _sqlClusterConnection: SqlClusterConnection;
+	private _sqlSession: sqlops.ObjectExplorerSession;
+	private _sqlConnProfile: sqlops.IConnectionProfile;
+	private _rootNode: SqlClusterRootNode;
 
-	public get uri(): string {
-		return this.sessionId || this._connection.uri;
+	constructor(arg: SqlClusterSessionArg) {
+		this._sqlClusterConnection = arg.sqlClusterConnection;
+		this._sqlSession = arg.sqlSession;
+		this._sqlConnProfile = arg.sqlConnProfile;
+		this._rootNode = new SqlClusterRootNode(this,
+			new TreeDataContext(arg.appContext.extensionContext, arg.changeHandler),
+			arg.sqlSession.rootNode.nodePath);
 	}
 
-	public get connection(): Connection {
-		return this._connection;
-	}
+	public get sqlClusterConnection(): SqlClusterConnection { return this._sqlClusterConnection; }
+	public get sqlSession(): sqlops.ObjectExplorerSession { return this._sqlSession; }
+	public get sqlConnProfile(): sqlops.IConnectionProfile { return this._sqlConnProfile; }
+	public get sessionId(): string { return this._sqlSession.sessionId; }
+	public get rootNode(): SqlClusterRootNode { return this._rootNode; }
 
-	public set root(node: RootNode) {
-		this._root = node;
+	public isMatchedSqlConnProfile(sqlConnProfile: sqlops.IConnectionProfile): boolean {
+		return this._sqlConnProfile.id === sqlConnProfile.id;
 	}
+}
 
-	public get root(): RootNode {
-		return this._root;
-	}
+interface SqlClusterSessionArg {
+	sqlClusterConnection: SqlClusterConnection;
+	sqlSession: sqlops.ObjectExplorerSession;
+	sqlConnProfile: sqlops.IConnectionProfile;
+	appContext: AppContext;
+	changeHandler: ITreeChangeHandler;
 }
 
-class RootNode extends TreeNode {
-	private children: TreeNode[];
-	constructor(private _session: Session, private context: TreeDataContext, private nodePath: string) {
+class SqlClusterRootNode extends TreeNode {
+	private _children: TreeNode[];
+	private _sqlClusterSession: SqlClusterSession;
+	private _treeDataContext: TreeDataContext;
+	private _nodePath: string;
+
+	constructor(sqlClusterSession: SqlClusterSession, treeDataContext: TreeDataContext, nodePath: string) {
 		super();
+		this._sqlClusterSession = sqlClusterSession;
+		this._treeDataContext = treeDataContext;
+		this._nodePath = nodePath;
 	}
 
-	public get session(): Session {
-		return this._session;
+	public get sqlClusterSession(): SqlClusterSession {
+		return this._sqlClusterSession;
 	}
 
-	public get nodePathValue(): string {
-		return this.nodePath;
+	public get nodePath(): string {
```

Comment:
As mentioned elsewhere, this is part of a path, so `nodePathValue` makes more sense

Reply:
fixed already.